File: src/main/search-provider-store.js
'use strict';

const fs = require('fs');
const path = require('path');
const crypto = require('crypto');

// %s is the query placeholder, replaced with the encoded search terms -
// the same convention most browsers use for custom search engine URLs.
const DEFAULT_PROVIDERS = [
  { id: 'duckduckgo', name: 'DuckDuckGo', urlTemplate: 'https://duckduckgo.com/?q=%s', builtIn: true },
  { id: 'startpage', name: 'Startpage', urlTemplate: 'https://www.startpage.com/sp/search?query=%s', builtIn: true },
  { id: 'brave', name: 'Brave Search', urlTemplate: 'https://search.brave.com/search?q=%s', builtIn: true },
  { id: 'mojeek', name: 'Mojeek', urlTemplate: 'https://www.mojeek.com/search?q=%s', builtIn: true },
];

/**
 * Lets the user choose which search engine the address bar submits
 * non-URL input to, and add their own (any https:// URL template with a
 * %s query placeholder). DuckDuckGo remains the default, matching prior
 * behavior when this didn't exist as a setting.
 */
class SearchProviderStore {
  constructor(userDataDir) {
    this.filePath = path.join(userDataDir, 'search-providers.json');
    const loaded = this._load();
    this.providers = loaded.providers;
    this.activeId = loaded.activeId;
  }

  _load() {
    if (fs.existsSync(this.filePath)) {
      try {
        const data = JSON.parse(fs.readFileSync(this.filePath, 'utf8'));
        if (Array.isArray(data.providers) && data.providers.length > 0) {
          return { providers: data.providers, activeId: data.activeId || data.providers[0].id };
        }
      } catch (err) {
        // fall through to defaults
      }
    }
    return { providers: DEFAULT_PROVIDERS.map((p) => ({ ...p })), activeId: 'duckduckgo' };
  }

  _persist() {
    fs.writeFileSync(this.filePath, JSON.stringify({ providers: this.providers, activeId: this.activeId }, null, 2));
  }

  list() {
    return this.providers.map((p) => ({ ...p, active: p.id === this.activeId }));
  }

  getActive() {
    return this.providers.find((p) => p.id === this.activeId) || this.providers[0];
  }

  setActive(id) {
    if (!this.providers.some((p) => p.id === id)) return { ok: false, reason: 'not-found' };
    this.activeId = id;
    this._persist();
    return { ok: true };
  }

  add(name, urlTemplate) {
    if (!name || !urlTemplate) return { ok: false, reason: 'Name and URL template are required.' };
    if (!urlTemplate.includes('%s')) return { ok: false, reason: 'URL template must include %s as the query placeholder.' };
    try {
      const parsed = new URL(urlTemplate.replace('%s', 'test'));
      if (parsed.protocol !== 'https:') return { ok: false, reason: 'Only https:// URLs are supported.' };
    } catch (err) {
      return { ok: false, reason: 'Invalid URL template.' };
    }
    const id = crypto.randomUUID();
    this.providers.push({ id, name, urlTemplate, builtIn: false });
    this._persist();
    return { ok: true, id };
  }

  remove(id) {
    const provider = this.providers.find((p) => p.id === id);
    if (!provider) return { ok: false, reason: 'not-found' };
    if (provider.builtIn) return { ok: false, reason: 'Built-in providers cannot be removed.' };
    this.providers = this.providers.filter((p) => p.id !== id);
    if (this.activeId === id) this.activeId = this.providers[0]?.id || 'duckduckgo';
    this._persist();
    return { ok: true };
  }

  /** Builds the search URL for a query using the currently active provider. */
  buildSearchUrl(query) {
    const provider = this.getActive();
    return provider.urlTemplate.replace('%s', encodeURIComponent(query));
  }
}

module.exports = { SearchProviderStore, DEFAULT_PROVIDERS };

File: android/app/src/main/java/org/dago/browser/SearchProviderStore.kt
package org.dago.browser

import android.content.Context
import org.json.JSONArray
import org.json.JSONObject
import java.net.URLEncoder
import java.util.UUID

/**
 * Android counterpart to the desktop app's search-provider-store.js: lets
 * the user choose which search engine the address bar submits non-URL
 * input to, and add their own (any https:// URL template with a %s query
 * placeholder). DuckDuckGo remains the default.
 */
class SearchProviderStore(context: Context) {
    data class Provider(val id: String, val name: String, val urlTemplate: String, val builtIn: Boolean, val active: Boolean)

    companion object {
        private val DEFAULT_PROVIDERS = listOf(
            Triple("duckduckgo", "DuckDuckGo", "https://duckduckgo.com/?q=%s"),
            Triple("startpage", "Startpage", "https://www.startpage.com/sp/search?query=%s"),
            Triple("brave", "Brave Search", "https://search.brave.com/search?q=%s"),
            Triple("mojeek", "Mojeek", "https://www.mojeek.com/search?q=%s"),
        )
    }

    private val prefs = context.getSharedPreferences("dago_search_providers", Context.MODE_PRIVATE)

    init {
        if (!prefs.contains("providers")) {
            val defaults = JSONArray()
            for ((id, name, url) in DEFAULT_PROVIDERS) {
                defaults.put(JSONObject().put("id", id).put("name", name).put("urlTemplate", url).put("builtIn", true))
            }
            prefs.edit().putString("providers", defaults.toString()).putString("activeId", "duckduckgo").apply()
        }
    }

    fun list(): List<Provider> {
        val raw = loadRaw()
        val activeId = prefs.getString("activeId", "duckduckgo")
        val result = mutableListOf<Provider>()
        for (i in 0 until raw.length()) {
            val obj = raw.getJSONObject(i)
            val id = obj.getString("id")
            result.add(Provider(id, obj.getString("name"), obj.getString("urlTemplate"), obj.getBoolean("builtIn"), id == activeId))
        }
        return result
    }

    fun getActive(): Provider = list().find { it.active } ?: list().first()

    fun setActive(id: String): Boolean {
        if (loadRaw().let { raw -> (0 until raw.length()).none { raw.getJSONObject(it).getString("id") == id } }) return false
        prefs.edit().putString("activeId", id).apply()
        return true
    }

    fun add(name: String, urlTemplate: String): Result<String> {
        if (name.isBlank() || urlTemplate.isBlank()) return Result.failure(IllegalArgumentException("Name and URL template are required."))
        if (!urlTemplate.contains("%s")) return Result.failure(IllegalArgumentException("URL template must include %s as the query placeholder."))
        if (!urlTemplate.startsWith("https://")) return Result.failure(IllegalArgumentException("Only https:// URLs are supported."))

        val raw = loadRaw()
        val id = UUID.randomUUID().toString()
        raw.put(JSONObject().put("id", id).put("name", name).put("urlTemplate", urlTemplate).put("builtIn", false))
        persist(raw)
        return Result.success(id)
    }

    fun remove(id: String): Boolean {
        val raw = loadRaw()
        var found = false
        val rebuilt = JSONArray()
        for (i in 0 until raw.length()) {
            val obj = raw.getJSONObject(i)
            if (obj.getString("id") == id) {
                if (obj.getBoolean("builtIn")) return false // cannot remove built-in providers
                found = true
            } else {
                rebuilt.put(obj)
            }
        }
        if (!found) return false
        persist(rebuilt)
        if (prefs.getString("activeId", null) == id) {
            val fallback = if (rebuilt.length() > 0) rebuilt.getJSONObject(0).getString("id") else "duckduckgo"
            prefs.edit().putString("activeId", fallback).apply()
        }
        return true
    }

    /** Builds the search URL for a query using the currently active provider. */
    fun buildSearchUrl(query: String): String {
        val encoded = URLEncoder.encode(query, "UTF-8")
        return getActive().urlTemplate.replace("%s", encoded)
    }

    private fun loadRaw(): JSONArray {
        val raw = prefs.getString("providers", null) ?: return JSONArray()
        return runCatching { JSONArray(raw) }.getOrDefault(JSONArray())
    }

    private fun persist(providers: JSONArray) {
        prefs.edit().putString("providers", providers.toString()).apply()
    }
}
